Refactor ClientCertAuthChallengeHandler

melissaahn opened the pull request:
_Note: Updated 9/22 to reflect changes made during review of this PR._
## Summary
YubiKit was recently added as a 3P library to Common in order to interact with YubiKeys for smartcard CBA. Currently, YubiKit methods are directly being called within `ClientCertAuthChallengeHandler`, which functions for the purposes of meeting Executive Order requirements.

However, refactoring `ClientCertAuthChallengeHandler` to use abstract methods to communicate with smartcards is better in practice, as it will:
- allow the creation of automatic testing for this feature to be completed more easily (no need to interact directly with YubiKit methods for testing dialogs).
- contain the YubiKit references to specific classes.
- allow us to more easily add support for other smartcard brands when we need to.

This PR decouples YubiKit from `ClientCertAuthChallengeHandler` by adding new abstract classes that facilitate smartcard CBA:
- `ICertDetails`: Must hold an `X509Certificate`, but should hold anything specifically needed to identify a certificate on the smartcard.
- `AbstractSmartcardCertBasedAuthManager`: this replaces YubiKitManager within `ClientCertAuthChallengeHandler` and handles usb discovery and creating `ISmartcardSession`s.
- `ISmartcardSession`: An abstract session object that implements methods that are more involved with the smartcard (and thus usually have more restrictions, which is why this is separate from the manager). 
- `CertBasedAuthFactory`: creates and holds an instance of a `ISmartcardCertBasedAuthManager` and `DialogHolder` for `ClientCertAuthChallengeHandler`. (More on this class later).

Implementation classes that are YubiKit specific were created and contain essentially the same logic that was in `ClientCertAuthChallengeHandler`. 

But wait... **there's more!**

With the help of initial feedback for this PR, `ClientCertAuthChallengeHandler` is split into two classes: `SmartcardCertBasedAuthChallengeHandler` and `OnDeviceCertBasedAuthChallengeHandler`. As the names suggest, the two classes handle smartcard CBA and on device CBA separately. An instance of `CertBasedAuthFactory` is injected into `AzureActivityDirectoryWebViewClient`, and it helps create and return an appropriate `ICertBasedAuthChallengeHandler` in the `onReceivedClientCertRequest` method.
With the addition of these two classes, `ClientCertAuthChallengeHandler` will be deleted.

## Testing
I ensured that the basic success scenarios and failure scenarios outlined in our test plans are still producing the same results, and I will work on some automated testing for more error scenarios (see Next Steps below).

## Next Steps
I plan to work on automatic testing while this PR is being reviewed. ~~I think creating a shadow of the Factory which instead returns a test implementation of `ISmartcardCertBasedAuthManager` is the easiest way to go, but let me know if there is a better way to do this in practice.~~ _Edit: Going to create test implementations of `AbstractSmartcardCertBasedAuthManager` and `ISmartcardSession` and inject them into the constructor of `CertBasedAuthFactory`.

I also think that when we do have to add more support for other smartcard brands, `CertBasedAuthFactory` will instead hold a List of `ISmartcardCertBasedAuthManager`s (since we won't know which brand of smartcard the user will have until they plug it in) and will be altered to iterate through this list to determine which type of device is connected.

## Related PRs
- Merged the smartcard CBA feature into dev, contains a long description about the feature: https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/1814

melissaahn committed:
checkpoint

melissaahn committed:
saving progress

melissaahn committed:
Merge remote-tracking branch 'origin' into melissaahn/CertBasedAuthManager

melissaahn committed:
functional

melissaahn committed:
some clean up in manager and challengehandler

melissaahn committed:
added javadocs and separated sessions

melissaahn committed:
extracted some methods for logging

melissaahn commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -214,168 +105,92 @@ public void run() {
+            return null;
         }
         //Else, proceed with user certificates stored on device.
         handleOnDeviceCertAuth(request);
```
I think it was brought up in my previous PR that maybe regular on-device CBA (which is started in the method `handleOnDeviceCertAuth`) and smartcard CBA should be separate features in their own classes. Are there any thoughts on how this could be done to improve the structure of `ClientCertAuthChallengeHandler`?

iambmelt replied:
I would suggest pulling these capabilities into separate interfaces and having the functions/methods which require this behavior accept an instance of that interface, rather than object that can do both

iambmelt replied:
(If a given class can do both, it should implement both of those interfaces)

melissaahn replied:
If I understand correctly, this would mean creating an interface for on-device CBA and a separate one for smartcard CBA, and then feeding in implementations of these interfaces to `ClientCertAuthChallengeHandler` as parameters? This would leave `ClientCertAuthChallengeHandler` without the actual implementations of the features, so it can just call the interfaces to handle CBA as necessary.

iambmelt replied:
I might have given clumsy feedback in my initial review.... trying again.

My.... far at a distance.... view would be that "CBA is CBA".... sometimes it happens on device, and sometimes it happens via smartcard

You might want a generic "Can-perform-CBA" interface and have different implementations of that interface which know how to do it on-device and via smartcard. Then.... pass the appropriate instance at runtime based on which you're doing.

If they accept/need different params, you can hide this detail in a parameters object that can work across either class.

So to summarize with a pseudocode example....

```java
public interface ICertBasedAuthHandler {
    void performCba(ClientCertRequest request);
}

public class OnDeviceCbaHandler implements ICertBasedAuthHandler {
    @Override
    void performCba(ClientCertRequest request) {
         // provide implementation
    }
}

public class SmartCardCbaHandler implements ICertBasedAuthHandler {
    @Override
    void performCba(ClientCertRequest request) {
         // provide implementation
    }
}

// Then.... hide the if/else behind object construction... A factory can help here....
public static ICertBasedAuthHandler createCertBasedAuthHandler(boolean useSmartCard) {
    return useSmartCard? new SmartCardCbaHandler() : new OnDeviceCbaHandler();
}

// Put it all together and your method starts to look like this....
@RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
@Override
public Void processChallenge(@NonNull final ClientCertRequest request) {
    final ICertBasedAuthHandler cbaHandler = createCertBasedAuthHandler(mSmartcardCertBasedAuthManager.isDeviceConnected());
    cbaHandler.performCba(request);
    return null;
}
```

melissaahn commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/SmartcardCertBasedAuthManagerFactory.java:
```diff
@@ -0,0 +1,42 @@
+    public static ISmartcardCertBasedAuthManager getSmartcardCertBasedAuthManager(Activity activity) {
+        //Return instance of YubiKitCertBasedAuthManager, since this is the only implementation of
+        // ISmartcardCertBasedAuthManager we have right now.
+        return new YubiKitCertBasedAuthManager(activity);
```
Right now, this only returns the YubiKit manager implementation, but I can imagine in the future that this method will instantiate multiple managers and return a list. 
~~For testing, I wasn't sure of a way to differentiate between a testing scenario (where we would need a test implementation of the manager interface) and an actual user scenario based on the Activity, so my plan is to create a shadow of `SmartcardCertBasedAuthManagerFactory` and return a test implementation the manager instead... is this ok in practice?~~_Edit: Adding an extra parameter to ClientCertAuthChallengeHandler contructor that allows us to pass in the ISmartcardCertBasedAuthManager. For testing, I'll pass a test implementation of ISmartcardCertBasedAuthManager._

melissaahn committed:
Merge branch 'dev' into melissaahn/CertBasedAuthManager

melissaahn committed:
fixed lock errors

melissaahn committed:
Merge remote-tracking branch 'origin' into melissaahn/CertBasedAuthManager

melissaahn committed:
added extra parameter field to clientcertauthchallengehandler

melissaahn commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -46,166 +40,65 @@
      */
-    public ClientCertAuthChallengeHandler(@NonNull final Activity activity) {
+    public ClientCertAuthChallengeHandler(@NonNull final Activity activity,
+                                          @NonNull final ISmartcardCertBasedAuthManager smartcardCertBasedAuthManager) {
```
Want to highlight change here: directly passing `ISmartcardCertBasedAuthManager` in constructor.

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -46,166 +40,65 @@
+     * @param smartcardCertBasedAuthManager ISmartcardCertBasedAuthManager instance.
      */
-    public ClientCertAuthChallengeHandler(@NonNull final Activity activity) {
+    public ClientCertAuthChallengeHandler(@NonNull final Activity activity,
```
Do you strictly need an `Activity` here? Or would a `Context` suffice?

Can this be called from a `Fragment`?

melissaahn replied:
I think in this case, we do need the host `Activity` because an `Activity` is required as a parameter for [`KeyChain.choosePrivateKeyAlias`](https://developer.android.com/reference/android/security/KeyChain#choosePrivateKeyAlias(android.app.Activity,%20android.security.KeyChainAliasCallback,%20java.lang.String[],%20java.security.Principal[],%20java.lang.String,%20int,%20java.lang.String)), which we call for regular on-device CBA.
Though when I do end up creating separate interfaces for on-device and smartcard CBA, the `Activity` parameter will be removed, instead being passed into the interface for on-device CBA.

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/WebViewAuthorizationFragment.java:
```diff
@@ -261,15 +261,15 @@ public boolean onTouch(final View view, final MotionEvent event) {
+            mAADWebViewClient.stopSmartcardUsbDiscovery();
         } else {
-            Logger.error(methodTag, "YubiKitManager usb discovery not stopped due to mAADWebViewClient being null", null);
+            Logger.error(methodTag, "Usb discovery not stopped due to mAADWebViewClient being null", null);
```
Error message exposes the name of a private variable. Would suggest something more meaningful for somebody reviewing the log who may not have source access

melissaahn replied:
This was changed to, "Fragment destroyed, but smartcard usb discovery was unable to be stopped."

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/AzureActiveDirectoryWebViewClient.java:
```diff
@@ -85,8 +86,9 @@ public AzureActiveDirectoryWebViewClient(@NonNull final Activity activity,
-        //Creating ClientCertAuthChallengeHandler starts YubiKitManager usb discovery
-        mClientCertAuthChallengeHandler = new ClientCertAuthChallengeHandler(getActivity());
+        //Creating ClientCertAuthChallengeHandler starts smartcard usb discovery
+        mClientCertAuthChallengeHandler = new ClientCertAuthChallengeHandler(getActivity(),
```
Calling a constructor in a constructor is _usually_ a sign your class needs to be accepting its dependency, rather than creating it. Consider passing in the instance of `ClientCertAuthChallengeHandler` or supplying an interface for it. This can make these classes easer to test separately from each other

rpdome replied:
https://java-design-patterns.com/patterns/dependency-injection/

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/AzureActiveDirectoryWebViewClient.java:
```diff
@@ -466,14 +468,14 @@ public void onReceivedClientCertRequest(WebView view,
+            mClientCertAuthChallengeHandler.stopSmartcardUsbDiscovery();
         } else {
-            Logger.error(methodTag, "YubiKitManager usb discovery not stopped due to mClientCertAuthChallengeHandler being null", null);
+            Logger.error(methodTag, "Usb discovery not stopped due to mClientCertAuthChallengeHandler being null", null);
```
Another call out for a private variable name in a log message

melissaahn replied:
I ended up deleting this log since `mClientCertAuthChallengeHandler` is set to a `@Nonnull` constructor parameter.

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -214,168 +107,92 @@ public void run() {
-        //A connection to the YubiKey needs to be made in order to read the certificates off it.
-        getActivePivSessionAsync(request, new IPivSessionCallback() {
+
+        mSmartcardCertBasedAuthManager.attemptDeviceSession(new ISmartcardCertBasedAuthManager.ISessionCallback() {
```
This is a bit of a subjective style-nit, so take with a pinch of salt.

`attemptDeviceSession` sounds like the name of method that doesn't think it will succeed..... and maybe that's true, but the name doesn't inspire confidence in the caller.

When these kinds of "maybe this will work!" methods crop-up, I prefer to make the compiler do the talking.

For instance, if the method `throws`, the caller must catch the `Exception` thereby signaling that it can fail.

Alternatively, if the method accepts an `interface` with an error callback, we also know that it can fail.

melissaahn replied:
Makes sense... renaming to `requestDeviceSession`.

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -214,168 +107,92 @@ public void run() {
-                } catch(final IOException | ApduException | BadResponseException e) {
-                    Logger.error(methodTag, e.getMessage(), e);
-                    //Show general error dialog.
+            public void onGetSession(@NonNull ISmartcardSession session) throws Exception {
```
```suggestion
            public void onGetSession(@NonNull final ISmartcardSession session) throws Exception {
```

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -214,168 +107,92 @@ public void run() {
+                    return;
+                }
+                //Create List that contains cert details only pertinent to the cert picker.
+                final List<ICertDetails> certList = session.getCertDetailsList();
```
Can this ever be `null`? Do we need to check?

melissaahn replied:
`getCertDetailsList()` should never return null (method has `@NonNull` annotation). In the case of no certs on smartcard, an empty list is returned.

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -448,108 +257,60 @@ public void onGetSession(@NonNull PivSession piv) {
+                                         @NonNull final ISmartcardSession session,
+                                         @NonNull final ClientCertRequest request)
+            throws Exception {
+        //final String methodTag = TAG + "useSmartcardCertForAuth:";
```
Can be cleaned up?

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ICertDetails.java:
```diff
@@ -0,0 +1,41 @@
+import java.security.cert.X509Certificate;
+
+/**
+ * Holds X509Certificate and other important details that may be smartcard brand dependent.
```
```suggestion
 * Holds X509Certificate and other important details that may be implementation specific.
```

Suggesting since we don't need to artificially constraint this interface's use to smartcards

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ISmartcardCertBasedAuthManager.java:
```diff
@@ -0,0 +1,90 @@
+    /**
+     * Callback methods to be run upon initial connection and disconnection of a smartcard device.
+     */
+    interface IStartDiscoveryCallback {
```
Can this ever error out? Should there be an error handling callback here too?

(Even if Yubikey APIs won't throw an error, what about another OEM?)

melissaahn replied:
Could be possible with other smartcard APIs we may use later. I'm adding a `onException` method to the object.

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/SmartcardCertBasedAuthManagerFactory.java:
```diff
@@ -0,0 +1,42 @@
+    public static ISmartcardCertBasedAuthManager getSmartcardCertBasedAuthManager(Activity activity) {
+        //Return instance of YubiKitCertBasedAuthManager, since this is the only implementation of
+        // ISmartcardCertBasedAuthManager we have right now.
+        return new YubiKitCertBasedAuthManager(activity);
```
Do you really need an `Activity`? Or would a `Context` suffice?

Consider portability with:
- Headless IoT devices
- Use from a `Fragment`

melissaahn replied:
As you pointed out below, all I need is the context :)

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/YubiKitCertBasedAuthManager.java:
```diff
@@ -0,0 +1,220 @@
+     * @param activity current host activity.
+     */
+    YubiKitCertBasedAuthManager(Activity activity) {
+        mYubiKitManager = new YubiKitManager(activity.getApplicationContext());
```
👀

For principle of least privilege, consider accepting the least-specific class you need to fulfill this class' requirements. Looks like what you really is a `Context`, rather than the whole `Activity`

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/YubiKitCertBasedAuthManager.java:
```diff
@@ -0,0 +1,220 @@
+     * @param startDiscoveryCallback Contains callbacks to run when a YubiKey is connected and disconnected.
+     */
+    @Override
+    public void startDiscovery(IStartDiscoveryCallback startDiscoveryCallback) {
```
```suggestion
    public void startDiscovery(@NonNull final IStartDiscoveryCallback startDiscoveryCallback) {
```

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ISmartcardCertBasedAuthManager.java:
```diff
@@ -0,0 +1,90 @@
+    /**
+     * Runs smartcard brand specific processes that may need to occur prior to authentication.
+     */
+    void prepareForAuth();
```
Naming suggestion...

`initialize()`

melissaahn replied:
Addressed in the comment thread by `YubiKitCertBasedAuthManager` implementation of `prepareForAuth()`

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/YubiKitCertBasedAuthManager.java:
```diff
@@ -0,0 +1,220 @@
+     * Adds a PivProvider instance to the Java static Security List (and emits relevant telemetry).
+     */
+    @Override
+    public void prepareForAuth() {
```
Suggested above -- this method naming seems a bit assumptive about the use of the smartcard. What if we're just signing something using its cert? Is that auth?

Suggest naming this more generically, along the lines of `initialize()` or similar

melissaahn replied:
I created `PrepareForAuth` because some logic needed to happen (specifically, adding an implemented security provider to the static Java Security list) **right before** `proceed` is called on the `ClientCertRequest`.  I'm not certain that `initialize()` would be specific enough here since I wouldn't want someone to think it should be called any time other than before `proceed`. Would `initializeBeforeProceedingWithRequest()` be an ok rename? It's kinda long, but at least more accurate than `prepareForAuth` :)

melissaahn replied:
Ended up renaming to `initBeforeProceedingWithRequest`. Chatted with Brian offline regarding the purpose of this method, and after thinking about it for a bit, I think it could still be useful for future implementations, as long as the developer makes sure to clean up well in `onDestroy` or the `onClosedConnection` callback. 
I also think it's more organized to separate the PivProvider logic from the `getKeyForAuth` method in order to avoid hiding the PivProvider logic in a semi related method. 
I'll probably revisit this in the future, especially when we add support for more types of smartcards.

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/YubiKitSmartcardSession.java:
```diff
@@ -0,0 +1,174 @@
+     * Creates instance of YubiKitSmartcardSession.
+     * @param p PivSession created.
+     */
+    public YubiKitSmartcardSession(PivSession p) {
```
`NonNull`? `@Nullable`?

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/YubiKitSmartcardSession.java:
```diff
@@ -0,0 +1,174 @@
+            certList.add(new YubiKitCertDetails(cert, slot));
+        } catch (final ApduException e) {
+            //If sw is 0x6a82 (27266), This is a FILE_NOT_FOUND error, which we should ignore since this means the slot is merely empty.
+            if (e.getSw() == 0x6a82) {
```
Oooh magic numbers

Let's pull this into a `final static` const and document it.

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/YubiKitSmartcardSession.java:
```diff
@@ -0,0 +1,174 @@
+     * @throws IOException in case of connection error
+     */
+    @Override
+    public boolean verifyPin(char[] pin) throws ApduException, IOException {
```
Not seeing any `null` check here, not sure if it can be -- maybe you want to be defensive on sanitizing the input

melissaahn replied:
Adding a null check for a EditText component that could lead to an error, and verified that `pin` is `@Nonnull`

rpdome replied:
I think what @iambmelt meant is that you might want to add an extra check here that log + return/throw an error if the array is null or empty. (is it valid to have an empty pin)?

melissaahn replied:
I followed the path to get to the pin array here, which included finding that EditText could be null (and handling the null check at that point), and I _think_ now the Pin array shouldn't ever be null. 
Right now there isn't specific logic to treat an empty pin array any differently than a filled one. I'm not sure if it's possible to actually set an empty PIN, but I'm thinking it's probably highly discouraged :) 
An enhancement I can make in a future PR is to disable the pin AlertDialog positive button when the EditText is empty so that the user can never submit an empty PIN.

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/YubiKitSmartcardSession.java:
```diff
@@ -0,0 +1,174 @@
+     */
+    @NonNull
+    @Override
+    public PrivateKey getKeyForAuth(ICertDetails certDetails, char[] pin) throws Exception {
```
Final? Null check on input?

melissaahn replied:
Since `ArrayAdapter.getItem()` could return null, I added a null check after that method is called within `SmartcardCertPickerDialog`. Made parameters in `getKeyForAuth` `@Nonnull`.

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/AzureActiveDirectoryWebViewClient.java:
```diff
@@ -85,8 +86,9 @@ public AzureActiveDirectoryWebViewClient(@NonNull final Activity activity,
-        mClientCertAuthChallengeHandler = new ClientCertAuthChallengeHandler(getActivity());
+        //Creating ClientCertAuthChallengeHandler starts smartcard usb discovery
+        mClientCertAuthChallengeHandler = new ClientCertAuthChallengeHandler(getActivity(),
+                SmartcardCertBasedAuthManagerFactory.getSmartcardCertBasedAuthManager(getActivity()));
```
Nit on naming

`get` here in a static context connotes a Singleton like functionality imo.... if it's not breaking, I might suggest a factory method name that implies object construction like `createSmartcardCertBasedAuthManager()`

melissaahn replied:
Renamed to `createSmartcardCertBasedAuthManager()`

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ISmartcardSession.java:
```diff
@@ -0,0 +1,67 @@
+     * @throws Exception that could vary upon smartcard type.
+     */
+    @NonNull
+    PrivateKey getKeyForAuth(ICertDetails certDetails, char[] pin) throws Exception;
```
Does the `PrivateKey` enter the apps memory here? That seems kinda weird.... I would have thought it was not exposed to the app itself, like a TPM behaves

melissaahn replied:
From my understanding, the `PrivateKey` being obtained from YubiKit (`PivPrivateKey`, in their case) isn't an actual private key, but there are methods within the object which allow it to perform private key operations without removing the private key from the smartcard. YubiKit includes SPI implementations that call these methods after `proceed` is called on `ClientCertRequest`.

melissaahn replied:
But I can see how the name `getKeyForAuth`, which returns a `PrivateKey`, can give off the wrong impression here.... [`proceed`](https://developer.android.com/reference/android/webkit/ClientCertRequest#proceed(java.security.PrivateKey,%20java.security.cert.X509Certificate[])) requires a `PrivateKey` object, but not necessarily a private key. Do you have any renaming suggestions?

rpdome replied:
@iambmelt The PrivateKey here doesn't contain the raw (encoded) key (if you try to use it with other non-YubiKit Crypto implementations (i.e. the BouncyCastle one), it would fail).

(PivPrivateKey actually contains rawSignOrDecrypt() function)

iambmelt replied:
This all makes sense... Most TPMs have similar _looking_ interfaces where it appears that the `PrivateKey` enters memory but in reality what you get is an empty proxy object that behaves like a PrivateKey but doesn't introduce any key material into app memory space.

Comments here can help, but LGTM -- it makes sense as is

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/SmartcardCertBasedAuthManagerFactory.java:
```diff
@@ -0,0 +1,42 @@
+     * @param activity Current host activity.
+     * @return A ISmartcardCertBasedAuthManager implementation instance.
+     */
+    public static ISmartcardCertBasedAuthManager getSmartcardCertBasedAuthManager(Activity activity) {
```
Same feedback as above... `get` seems to imply this is a singleton/global object

melissaahn replied:
Renaming to `createSmartcardCertBasedAuthManager`

iambmelt reviewed (changes requested):
Left some suggested edits, feel free to reach out for clarifications. You'll need a signoff from another SME who's still actively contributing over here 😉 

Overall - love the nature of this changeset. It's great to see the coupling to Yubikey's implementation largely eliminated as part of this work

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/YubiKitSmartcardSession.java:
```diff
@@ -0,0 +1,174 @@
+            //If no InvalidPinException is thrown, PIN is validated.
+            return true;
+        } catch (InvalidPinException e) {
+            return false;
```
This looks like an event worth logging

melissaahn replied:
Added info log with message "Incorrect PIN entered."

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/AzureActiveDirectoryWebViewClient.java:
```diff
@@ -466,14 +468,14 @@ public void onReceivedClientCertRequest(WebView view,
+        final String methodTag = TAG + ":stopSmartcardUsbDiscovery";
         if (mClientCertAuthChallengeHandler != null) {
-            mClientCertAuthChallengeHandler.stopYubiKitManagerUsbDiscovery();
+            mClientCertAuthChallengeHandler.stopSmartcardUsbDiscovery();
```
I'd suggest adding onDestroy() to the IChallengeHandler<> interface, and then invoke stopSmartCardUspDiscovery() there.

That way, the implementation detail won't be leaked.

rpdome replied:
(and in the outer layer, WebviewAuthorizationFragment.onDestroy() can invoke IChallengeHandler.onDestroy() directly.

melissaahn replied:
Since `WebviewAuthorizationFragment` holds `AADWebViewClient` and not `IChallengeHandler`, I have it call `AADWebViewClient.onDestroy()` (I didn't add `onDestroy()` to the WebClient interface, since there are so many, but let me know if it would be necessary).

rpdome replied:
meant to say WebviewAuthorizationFragment -> AADWebViewClient -> IChallengeHandler. 😅

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -214,168 +107,92 @@ public void run() {
-                handleSmartcardCertAuth(request);
-                return null;
-            }
+        if (mSmartcardCertBasedAuthManager.isDeviceConnected()) {
```
I'm wondering if we should create 2 subclasses of IChallengeHandler, one for regular CBA and another for smartcard CBA.

Then create a factory method that determines the state (could be by invoking the line above), and return the right IChallengeHandler to AzureActiveDirectoryWebViewClient's constructor (or its caller).

melissaahn replied:
I'm not sure about providing a challenge handler to `AzureActiveDirectoryWebViewClient`'s constructor because at the time of `AzureActiveDirectoryWebViewClient`'s instantiation, we don't necessarily know if we're going with smartcard CBA or on-device yet. We have to check right when we receive the `ClientCertRequest`. 
If I write out a way this could work (for my understanding):
-  `mSmartcardCertBasedAuthManager` should be injected into `AzureActiveDirectoryWebViewClient`, and `AzureActiveDirectoryWebViewClient` will need to take care of starting usb discovery (instead of the challenge handler). 
- In `AzureActiveDirectoryWebViewClient.onReceivedClientCertRequest`, the Factory takes in `mSmartcardCertBasedAuthManager` as a parameter for its new method (we'lll call it `isDeviceConnected`) and uses it to determine whether a smartcard is connected, then returns a challenge handler based on that (smartcard CBA challenge handler will need the manager to be injected). The `processChallenge` method of the challenge handler is then called.

rpdome replied:
- You could probably instantiate an IChallengeHandler object in onReceivedClientCertRequest()
- When that happens, then you can 

``` Java

final ISmartcardCertBasedAuthManager mYubikitManager;
final ISmartcardCertBasedAuthManager mSomeOtherManager;
...

// Initializes all the manager. give it time to do discovery.
void initialize(){
      mYubikitManager = new YubiKitCertBasedAuthManager(context);
      mYubikitManager.startDiscovery();
      
      mSomeOtherManager = ..... some other smartcard provider.
      mSomeOtherManager.startDiscovery();
}

IChallengeHandler getChallengeHandler(final Context context) {
        if (mYubikitManager.isDeviceConnected()){
              return new SmartcardCertBasedAuthChallengeHandler(mYubiKitManager);
        }
        
        if (mSomeOtherManager.isDeviceConnected()){
              return new OnDeviceCertBasedAuthChallengeHandler(mSomeOtherManager);
        }
        
        return new OnDeviceCertAuthManager();
}

```

rpdome replied:
The factory could look like something along this line.

``` Java

private ISmartcardCertBasedAuthManager mYubikitManager;
private ISmartcardCertBasedAuthManager mSomeOtherManager;
...

// Initializes all the managers. give them time to do discovery.
void initialize(final Context context){
      mYubikitManager = new YubiKitCertBasedAuthManager(context);
      mYubikitManager.startDiscovery();
      
      mSomeOtherManager = ..... some other smartcard provider.
      mSomeOtherManager.startDiscovery();
}

IChallengeHandler getChallengeHandler() {
      if (mYubikitManager.isDeviceConnected()){
            return new SmartcardCertBasedAuthChallengeHandler(mYubiKitManager);
      }
        
      if (mSomeOtherManager.isDeviceConnected()){
            return new OnDeviceCertBasedAuthChallengeHandler(mSomeOtherManager);
      }
        
      return new OnDeviceCertAuthManager();
}

void destroy() {
      mYubikitManager.stopDiscovery();
      mSomeOtherManager.stopDiscovery();
}

```

Then 
- In AzureActiveDirectoryWebViewClient's constructor, create a factory object and invoke initialize() (or make it part of the constructor)
- In AzureActiveDirectoryWebViewClient.onReceivedClientCertRequest(), invoke getChallengeHandler()

melissaahn replied:
`AzureActiveDirectoryWebViewClient` now has an instance of `ClientCertAuthFactory` that holds the manager and dialogholder. I ended up injecting the manager and dialogholder into the constructor. In the future, I think the manager parameter can turn into a `List<AbstractSmartcardCertBasedAuthManager>`, where the list contains a manager per smartcard that we support.

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -46,166 +40,65 @@
+    public ClientCertAuthChallengeHandler(@NonNull final Activity activity,
+                                          @NonNull final ISmartcardCertBasedAuthManager smartcardCertBasedAuthManager) {
         mActivity = activity;
         mDialogHolder = new DialogHolder(mActivity);
```
You could also extract an interface from DialogHolder, and pass a mock object (that doesn't pop any dialog, and would throw an exception if an unexpected dialog is displayed) in the constructor in tests.

(Alternatively, you could also use Robolectric, but abstracting out DialogHolder would allow you to get rid of mActivity in the CBA subclass completely, especially now that you've already extracted the yubikit logic out.).

melissaahn replied:
For testing, I wanted to make sure that `ClientCertAuthChallengeHandler` will bring the correct dialogs up, and I need to be able to interact with the dialogs (type in EditText and see error message disappear, for example), so wouldn't I want the original `DialogHolder` for testing too? Let me know if I'm misunderstanding.
It does make sense though to maybe inject a `DialogHolder` into a smartcard CBA implementation of `ClientCertAuthChallengeHandler` so that the challenge handler doesn't need any sort of activity/context variable.

rpdome replied:
I was referring to the unit/integration tests, in which dialogs cannot be popped up.

During UI/Automation (e2e) tests, then yes, you should use the original DialogHolder.

melissaahn replied:
I think I was able to create some unit tests with DialogHolder using ShadowAlertDialog (unless these would still be considered UI/Automation tests): https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/1839/files
It takes about 9-12 ish seconds to run, so if this can run faster as UI/Automation tests, I can create those in a separate PR, and then write separate unit tests with a different implementation of DialogHolder.

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ISmartcardCertBasedAuthManager.java:
```diff
@@ -0,0 +1,90 @@
+     * Logic to prepare an Android device to detect smartcards via usb.
+     * @param startDiscoveryCallback Contains callbacks to run when a smartcard is connected and disconnected.
+     */
+    void startDiscovery(final IStartDiscoveryCallback startDiscoveryCallback);
```
nit: nonnull.

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ISmartcardCertBasedAuthManager.java:
```diff
@@ -0,0 +1,90 @@
+     * Attempt to get an instance of a session in order to carry out methods specific to ISmartcardSession.
+     * @param callback Contains callbacks to run when a ISmartcardSession is successfully instantiated and when any exception is thrown due to a connection issue.
+     */
+    void attemptDeviceSession(@NonNull final ISessionCallback callback);
```
nit: tryGetDeviceSession or getDeviceSession?

melissaahn replied:
ended up going with `requestDeviceSession()`... yea or nay?

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/YubiKitCertBasedAuthManager.java:
```diff
@@ -0,0 +1,220 @@
+    private static final String MDEVICE_NULL_ERROR_MESSAGE = "Instance UsbYubiKitDevice variable (mDevice) is null.";
+    private static final String YUBIKEY_PROVIDER = "YKPiv";
+
+    private final YubiKitManager mYubiKitManager;
```
I assume the plan is to mock YubiKit classes with Robolectric?

melissaahn replied:
I tried to directly mock/shadow the YubiKit classes for testing before this PR... with this refactorization work, I thought it would be easier to just create implementations of `ISmartcardCertBasedAuthManager` and `ISmartcardSession` strictly for testing only, since that's what the shadow of`PivSession` was pretty much turning into. 
I use Robolectric to interact with the `AlertDialog`s and make sure that the correct ones are popping up.

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/YubiKitCertBasedAuthManager.java:
```diff
@@ -0,0 +1,220 @@
+        //Note that this provider is removed when the UsbYubiKeyDevice connection is closed.
+        final PivProviderStatusEvent pivProviderStatusEvent = new PivProviderStatusEvent();
+        if (Security.getProvider(YUBIKEY_PROVIDER) != null) {
+            //Remove existing PivProvider.
```
nit: remove redundant comment.

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/YubiKitCertBasedAuthManager.java:
```diff
@@ -0,0 +1,220 @@
+            Telemetry.emit(pivProviderStatusEvent.putIsExistingPivProviderPresent(true));
+            Logger.info(methodTag, "Existing PivProvider was present in Security static list.");
+        } else {
+            //This is expected behavior.
```
nit: remove redundant comment.

rpdome reviewed (commented):
Left some comments. Thanks @iambmelt for the guidance!

Will look again once all comments are addressed

melissaahn committed:
Part 1 of addressing comments

melissaahn committed:
addressing comments pt2

melissaahn committed:
addressing comments pt3

melissaahn commented:
Somes notes (mostly for myself) after addressing first iteration of comments:
- There will be a future PR that splits the logic for smartcard CBA and on-device CBA. It will probably involve creating two separate implementations of `IChallengeHandler`.
- There will be a future PR to disable positive button in the PIN dialog when the `EditText` is empty.
- Tried seeing if the `AlertDialog`s could be handled with a context instead of an activity, but this proved to not really work as cleanly or well as using an activity. This is fine, as I'm most likely going to be injecting an instance of `DialogHolder` into `ClientCertAuthChallengeHandler` anyway.

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/AzureActiveDirectoryWebViewClient.java:
```diff
@@ -457,37 +458,22 @@ private void returnError(final String errorCode, final String errorMessage) {
-        } else {
-            Logger.error(methodTag, "ClientCertRequest cannot be handled due to mClientCertAuthChallengeHandler being null.", null);
-        }
+        mClientCertAuthChallengeHandler.processChallenge(clientCertRequest);
```
This cleanup looks nice :)

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/SmartcardCertBasedAuthManagerFactory.java:
```diff
@@ -0,0 +1,42 @@
+     * @param context Current application context.
+     * @return A ISmartcardCertBasedAuthManager implementation instance.
+     */
+    public static ISmartcardCertBasedAuthManager createSmartcardCertBasedAuthManager(Context context) {
```
```suggestion
    public static ISmartcardCertBasedAuthManager createSmartcardCertBasedAuthManager(@NonNull final Context context) {
```

(This will need an `import` statement to compile correctly)

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/SmartcardCertPickerDialog.java:
```diff
@@ -25,6 +25,7 @@
 import android.app.Activity;
 import android.content.Context;
 import android.content.DialogInterface;
+import android.nfc.Tag;
```
👀

Is this being used?

melissaahn replied:
must have snuck in there while I was trying to log...

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/YubiKitCertBasedAuthManager.java:
```diff
@@ -0,0 +1,218 @@
+                    try {
+                        final SmartCardConnection c = value.getValue();
+                        final PivSession piv = new PivSession(c);
+                        YubiKitSmartcardSession session = new YubiKitSmartcardSession(piv);
```
```suggestion
                        final YubiKitSmartcardSession session = new YubiKitSmartcardSession(piv);
```

iambmelt reviewed (commented):
Additional comments

melissaahn committed:
addressing cleanup comments

melissaahn committed:
Merge remote-tracking branch 'origin/dev' into melissaahn/CertBasedAuthManager

melissaahn committed:
split ClientCertAuthChallengeHandler into two

melissaahn commented:
As advised, I split `ClientCertAuthChallengeHandler` into two classes: `OnDeviceCertBasedAuthChallengeHandler` and `SmartcardCertBasedAuthChallengeHandler`, both of which implement `ICertBasedAuthChallengeHandler` (which extends `IChallengeHandler`). I'll delete `ClientCertAuthChallengeHandler` right before I merge this PR, so that any conversations on that class can be found more easily. 

`AzureActiveDirectoryWebViewClient` now has three variables relating to CBA: `mSmartcardCertBasedAuthManager` (injected), `mDialogHolder` (injected), and `mCertBasedAuthChallengeHandler` (set within the `onReceivedClientCertRequest` method, so not `final`). 

 `mSmartcardCertBasedAuthManager` is needed in order to start usb discovery, which depends on an instance of `mDialogHolder` since certain dialogs need to be brought up upon an unexpected disconnection. The latest `mCertBasedAuthChallengeHandler` is saved for Telemetry purposes (we need a way to emit telemetry after a `ClientCertRequest` is proceeded with within the challenge handler. 

I removed the `onDestroy` methods from the `IChallengeHandler` instance since usb discovery now happens within the manager, so `onDestroy` was added to `ISmartcardCertBasedAuthManager` instead.

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/YubiKitCertBasedAuthManager.java:
```diff
@@ -0,0 +1,226 @@
+                                Telemetry.emit(pivProviderStatusEvent.putPivProviderRemoved(false));
+                                Logger.info(TAG, "An instance of PivProvider was not present in Security static list upon YubiKey device connection being closed.");
+                            }
+                            startDiscoveryCallback.onClosedConnection();
```
I'm wondering if we could get rid of IStartDiscoveryCallback.

- onCreateConnection() is not really doing anything useful at this moment i think. (It's dismissing a dialog, but given the current flow, there shouldn't be any dialog at that point, and we're killing the flow when the device is disconnected).

- you could turn onClosedConnection() into an event listener. 

i.e. add this code into ISmartcardCertBasedAuthManager

``` Java
void setOnDeviceDisconnectedCallback(ICallback callback) {
        mOnDisconnectedCallback = callback;
}
```

then in line 103, invoke mOnDisconnectedCallback if not null.

This way, once we start supporting multiple smartcard providers, we will not be drowned in chain of callbacks when we're invoking discovery logic.

(and with the factory, we'll need to wire in the dialog logic only with the "chosen one")

melissaahn replied:
Supposing we get rid of onCreateConnection(), this is what would happen in the "unplug early" scenario:
1. User unplugs in the middle of the dialog flow, so another error dialog pops up asking the user to plug in again.
2. User plugs in their key, and they 'ok' the permission dialog.
3. They still see the error dialog asking them to plug in. 

It's possible to still have a dialog left over upon a new connection when the previous connection ended unexpectedly like this. I think it does result in a better user experience to have the preexisting dialog dismissed, but if the messiness of having onCreateConnection() outweigh this, I'm open to getting rid of it.

IStartDiscoveryCallback also has an additional method onException() that was added in response to a suggestion in this PR. It's not seen in the YubiKitCertBasedAuthManager because YubiKit's discovery methods aren't throwing exceptions, but as Brian pointed out, other types of smartcards could.

I think what I'll do for now is create a variable to hold an instance of IStartDiscoveryCallback in ISmartcardCertBasedAuthManager, and have it set ~~within the constructor~~ _within a method named setStartDiscoveryCallback() (I'm probably going to rename this class)_. StartDiscovery() then will have no parameters. Let me know if that would work.

melissaahn replied:
As for the dialog logic, I think the one way I can see there being issues is if a user tried to tap another smartcard via NFC while another smartcard is currently plugged in. I may need to add some logic then that makes sure every other manager says no device is connected before continuing with a new connection.

rpdome replied:
I thought we would fail the flow if the user unplugs (as an attempt to simplify error handling logic). Isn't that the case?

Allowing users to switch devices mid flow... could be very complicated.

rpdome replied:
(If we are going to support NFC... 

Given the nature of it (not physically plugged in, so I assume the connection might not be as stable, and intermittent connection loss might happen as the device moves).....  we might need another dialog to let the user explicitly choose which cert/device  will be used to do auth, and then once selected, then we go all the way with that cert/device. )

melissaahn replied:
We do fail the cba flow immediately. Any time an error dialog is showing, the request has been cancelled and a server error message shows in the background. 
usb discovery is active for the duration of the webview fragment, so if a user wants to retry, they are able to replugin their key right away while seeing an error dialog. Then, they'll click the back button and click "use a certificate" to restart the cba flow.

melissaahn replied:
Addressing NFC, I was thinking that we could still build upon the 'already plugged in' work by making the criteria of choosing on-device (not having smartcard plugged in) be another dialog prompting the user to either tap their smartcard or plug it in (and then a button at the bottom if the user wants to goto on-device certs). I was hoping to get some cxp input on how they envisioned NFC to work, but I think at this point I'm just going to try a solution out and see what they think.
Since NFC discovery works a bit differently (probably will need to make the user tap twice for reading certs and then checking PIN), this will probably result in another larger feature PR. :^)

melissaahn committed:
CertBasedAuthFactory handles managers and DialogHolder

melissaahn committed:
Merge remote-tracking branch 'origin/dev' into melissaahn/CertBasedAuthManager

melissaahn committed:
addressed spotbugs

melissaahn committed:
Merge remote-tracking branch 'origin/dev' into melissaahn/CertBasedAuthManager

melissaahn committed:
renamed prepareForAuth

melissaahn committed:
Merge branch 'dev' into melissaahn/CertBasedAuthManager

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/YubiKitCertDetails.java:
```diff
@@ -0,0 +1,71 @@
+ * Holds certificate found on YubiKey and its corresponding slot.
+ */
+public class YubiKitCertDetails implements ICertDetails {
+    private final X509Certificate cert;
```
Nit on naming convention for member fields: `mCert` and `mSlot`

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/CertBasedAuthFactory.java:
```diff
@@ -0,0 +1,104 @@
+     */
+    public CertBasedAuthFactory(@NonNull final Activity activity,
+                                @NonNull final AbstractSmartcardCertBasedAuthManager smartcardCertBasedAuthManager,
+                                @NonNull final DialogHolder dialogHolder) {
```
Something about a factory class knowing how to dismiss/show errors on dialogs feels a bit like it 'knows too much'. Ideally, I think there would be some kind of handler in the middle which separates the construction/connection events from actual UI. Given that `SmarcardCertBasedAuthChallengeHandler` _also_ takes a dependency on `DialogHolder`, this might be a little tricky to refactor-out

melissaahn replied:
I see what you mean... the challenge handlers and the dialog holder are quite intertwined, so it will probably take another large refactoring PR to sort that out. I'll be reaching out to you again if I do end up doing this in the future ;)

rpdome replied:
Maybe we can do something like this?

```Java

...

class AzureActiveDirectoryWebViewClient {
    CertBasedAuthFactory mFactory;
    ICertBasedAuthChallengeHandler mChallengeHandler;

    public AzureActiveDirectoryWebViewClient(){
        mFactory = new CertBasedAuthFactory();
    }

    public void onReceivedClientCertRequest(@NonNull final WebView view,
                                            @NonNull final ClientCertRequest clientCertRequest) {
        if (mChallengeHandler != null ){
            // If there is any dialog going on in the current handler, 
            // we probably would want to dismiss it + unhook dialog callback, etc.
            mChallengeHandler.cleanup();
        }
        
        mChallengeHandler =  mFactory.getChallengeHandler();
        mChallengeHandler.processChallenge(clientCertRequest);
    }

    public void onDestroy() {
        mCertBasedAuthFactory.onDestroy();
    }

    public void emitTelemetryForCertBasedAuthResult(@NonNull final RawAuthorizationResult response) {
        if (mChallengeHandler != null){
            mChallengeHandler.emitTelemetryForCertBasedAuthResult(response);
        }
    }
}

....

class CertBasedAuthFactory {
    private final YubiKitSmartCardCertBasedAuthManager mYubiKitManager;
    private final Activity mActivity;

    public CertBasedAuthFactory(@NonNull final Activity activity){
        mActivity = activity;
        mYubiKitManager = new YubiKitCertBasedAuthManager(mActivity.getApplicationContext()); 
        mYubiKitManager.startDiscovery();
    }

    public ICertBasedAuthChallengeHandler createCertBasedAuthChallengeHandler() {
        if (mSmartcardCertBasedAuthManager.isDeviceConnected()) {
            return new SmartcardCertBasedAuthChallengeHandler(mYubiKitManager, new DialogHolder(mActivity));
        } 
        
        return new OnDeviceCertBasedAuthChallengeHandler(mActivity);
    }

    public void onDestroy() {
        mYubiKitManager.onDestroy();
    }
}
...

class SmartcardCertBasedAuthChallengeHandler {
     
    public SmartcardCertBasedAuthChallengeHandler(@NonNull final AbstractSmartcardCertBasedAuthManager mManager,
                                                  @NonNull final DialogHolder dialogHolder) {
                                                  
        mManager.setDiscoveryCallback(.....);                                            
    }
}
```

(I don't think we need to inject objects into the factory class. (Factory class is generally the "orchestrator", meaning that it is the one that holds all the dependencies, and injecting them into others.)

melissaahn replied:
Is the idea to set the `DiscoveryCallback` after discovery has already started? I think in the YubiKit case this might be ok, but that would be because nothing is currently calling the `onException` method of the `DiscoveryCallback`. A different implementation of `onStartDiscovery` may need the `onException` method if something goes wrong while starting up an implementation specific version of discovery

melissaahn replied:
unless, `onException` is separated from `DiscoveryCallback`...

melissaahn replied:
I ended up separating `onException` into its own interface named `IDiscoveryExceptionCallback`, and this is set right before `startDiscovery` is called. I think I implemented all the changes described in the sample above

iambmelt reviewed (approved):
Approving with comments -- overall lgtm. You _might_ consider how to decouple the factory class from the ui as that dependency seems a bit tightly coupled but I don't want to block on this.

Let's get @rpdome's review too -- I'm only consulting here 😄

melissaahn committed:
renaming variables in YubiKitCertDetails

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/OnDeviceCertBasedAuthChallengeHandler.java:
```diff
@@ -0,0 +1,154 @@
+     */
+    @Override
+    public Void processChallenge(ClientCertRequest request) {
+        handleOnDeviceCertAuth(request);
```
could probably merge this function with handleOnDeviceCertAuth()

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/SmartcardCertBasedAuthChallengeHandler.java:
```diff
@@ -0,0 +1,301 @@
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    @Override
+    public Void processChallenge(ClientCertRequest request) {
+        handleSmartcardCertAuth(request);
```
Could probably merge this function with handleOnDeviceCertAuth()

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -46,166 +40,74 @@
+ * Handles Certificate Based Authentication by means of certificates provisioned onto external smartcards or the mobile devices themselves.
+ * Note that CBA requires API >= 21 (ClientCertRequest class available with API >= 21)
  */
 public final class ClientCertAuthChallengeHandler implements IChallengeHandler<ClientCertRequest, Void> {
```
Do we still need this class?

melissaahn replied:
Seems like all conversations on this class have now been resolved, so I'm going to delete this class now.

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/OnDeviceCertBasedAuthChallengeHandler.java:
```diff
@@ -0,0 +1,154 @@
+
+        // When ADFS server sends null or empty issuers, we'll continue with cert prompt.
+        if (acceptableCertIssuers != null) {
+            for (Principal issuer : acceptableCertIssuers) {
```
nit: final

rpdome reviewed (commented):
Left some comments. almost there :)

melissaahn committed:
deleted ClientCertAuthChallengeHandler

melissaahn committed:
moved DialogHolder handling to challenge handler

melissaahn committed:
combined into processChallenge

melissaahn committed:
some final clean up methods

melissaahn committed:
remove Factory from constructor of WebViewClient

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/AzureActiveDirectoryWebViewClient.java:
```diff
@@ -455,38 +456,33 @@ private void returnError(final String errorCode, final String errorMessage) {
     @TargetApi(Build.VERSION_CODES.LOLLIPOP)
     @Override
-    public void onReceivedClientCertRequest(WebView view,
+    public void onReceivedClientCertRequest(final WebView view,
```
nit: nonnull final

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/AbstractSmartcardCertBasedAuthManager.java:
```diff
@@ -81,21 +96,18 @@ interface IDiscoveryCallback {
+    /**
+     * Callback method to be run upon an exception thrown during discovery start up.
+     */
+    interface IDiscoveryExceptionCallback {
```
It's probably okay keeping this part of IDiscoveryExceptionCallback. 
Given that this doesn't really do anything - yet. I don't see it as a blocker.

rpdome reviewed (approved):
Thanks for the hard work :)

melissaahn committed:
Merge remote-tracking branch 'origin/dev' into melissaahn/CertBasedAuthManager

melissaahn committed:
modified changelog and added FB modifier on exception callback:
